chore: Convert QueryAutoRefresh to TypeScript functional React component [sc-48362]

eric-briscoe opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->
Convert QueryAutoRefresh to TypeScript functional React component

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR converts QueryAutoRefresh from es6 class component to TypeScript functional component as part of bigger effort to migrate frontend code to TypeScript and functional React components defined in: https://github.com/apache/superset/discussions/18388

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
There should be no visual change to the application

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
1. Open SQL Lab / SQL Editor
2. Click the "RUN" button
3. Ensure Query runs and updates results with same behavior that existed prior to this PR
4. Test with both default behavior and asynchronous query execution on for a database

**Changes to http requests for getting Query statuses:**
If you watch the dev tools network tab you should see less network traffic as the execution now ensures only one outstanding http call to check query status is made at a time to avoid http response race conditions that can lead to incorrect application state and to reduce tandem network calls on slower connections.  If the call fails, the offline state change will be triggered but the interval timer will continue attempting to reconnect and set user online and show results once the connection is restored.  This can be tested by:
1. Opening the dev tools / Network / set the connection from "No Throttling" to "Offline", 2. Click the "Run" button as described in the initial test steps above
3. See the network failure and the UI will display an "Offline" pill (this is not new UI behavior)
4. Set the network back to "No Throttling" and wait a few seconds and the next interval will make a successful query status API call and you will see data results again.

QueryAutoRefresh was separated into two files to separate Redux knowledge from the simple functional component:
1. QueryAutoRefresh/index.ts - Simple functional component with no knowledge of Redux
3. QueryAutoRefresh/QueryAutoRefreshContainer.ts - Functional component container with connection to redux state that imports and contains QueryAutoRefresh/index.ts component

**IMPORTANT:** 
When changing QueryAutoRefresh to a functional component the behavior of setInterval no longer worked correctly.  This is a known issue with functional React components using setInterval can provide stale state / props in the interval callback creating undesired, hard to debug / fix behavior.  A new utility hook (SqlLab/utils/useInterval.ts) was introduced in this PR to handle this based on this very helpful article by Dan Abramov (thank you!): https://overreacted.io/making-setinterval-declarative-with-react-hooks/

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [X] Has associated issue: https://github.com/apache/superset/discussions/18388
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

eric-briscoe committed:
git commit -m 'Convert QueryAutoRefresh to functional component [sc-48362]'

eschutho commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,118 @@
+
+// returns true if the Query.state matches one of the specifc values indicating the query is still processing on server
+export const isQueryRunning = (q: Query): boolean =>
+  ['running', 'started', 'pending', 'fetching'].indexOf(q?.state) >= 0;
```
nit, but `['running', 'started', 'pending', 'fetching'].includes(q?.state)` may be just a bit more terse.

eschutho replied:
also, apologies, I just realized that this prob isn't your code, since it's a conversion to ts. :)

eric-briscoe replied:
Great catch, has been updated to use includes

eschutho commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,118 @@
+  // if there are started or running queries, this method should return true
+  const now = Date.now();
+  if (queryList && typeof queryList === 'object') {
+    shouldCheck = Object.values(queryList).some(
```
The typing in the function params for queryList is an array and not optional. If we want to rely on that, then do we need the second runtime check here for type? Asking only because if a js file is passing in the data, we're not guaranteed that the typing will match, so I'm wondering if that's the intent. We don't really have strong guidelines one way or the other, but it's a good conversation to have.

eric-briscoe replied:
@eschutho This is definitely a great area for discussion!  I usually try to make code runtime fault tolerant and verify that behavior in the unit tests.  Because TypeScript is only doing static code analysis at compile time it really provides not actual safety for the runtime execution.  TypeScript will catch a TON of places in code where the wrong type of value is passed, but it cannot guarantee null, undefined, invalid data ends up in redux or passed down on props at runtime.  The associated unit tests failed when I passed a null or undefined value so I added the explicit check if the value is both truthy and object because `typeof null === 'object'`.  

Also verifying the typeof is object helps avoid odd Object.values behavior if someone happened to pass a string instead of object this will make each character of a string a value in the response array.  This portion of code is not performance sensitive where it is executing rapidly in a loop, or expensive to do these checks so I felt it was worth the check to make the code bomb proof.  Pretty flexible on this if anyone has strong opinions otherwise but my go to is to try to ensure there is no way a function can produce an uncaught exception. 

React rendering loops can be pretty unforgiving where one call like object.values(null) in render loop can lead you to white screen of doom 😮 Trust but verify I always say :) 

Another approach I have used in the past is to add utility functions that can be imported like isArray, isObject, isString, etc so things like `if (queryList && typeof queryList === 'object') {` can be reduced to `if (isObject(queryList)) {`.  Having those utilities can lead to consistent code practice for validating inputs and help eliminate potential runtime errors.

eschutho replied:
Cool, yeah I agree with runtime checks, too. This sounds great. The only part that reads a little confusing to me, understanding that `typeof` an array is an object, but I suppose currently we're allowing an object here as well, esp since we take Object.values before running the array method on it. 

How do you feel about doing an `Array.isArray` check on queryList to be more specific on only allowing an array? And then I suppose we wouldn't need the Object.values, but could just do `queryList.some`. Does that still work here?

eric-briscoe replied:
@eschutho Your question here made me re-evaluate the code and realized that the confusion you are having is due to me marking the queryList arg as Query[] when it really is receiving a Dictionary of queries.  I have pushed a commit that resolves this mismatch and also am now using lodash's isObject since we are already using lodash in the project and it provides all the type checking utils like isObject, isString, etc and we do not need to reinvent the wheel.  Please take a look at the updated code and see if it makes more sense now.  It was not doing an Object.keys on an array, it was receiving an object setup as a dictionary with query id as the key, but I treated the arg like an array when adding the TypeScript annotations which was a mistake.  Nice catch!

eschutho replied:
Oh, ya, that makes more sense. 👍  :)

eschutho commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,118 @@
+
+  useEffect(() => {
+    if (isOffline !== offline) {
+      actions?.setUserOffline(isOffline);
```
Is there a benefit to keeping the local state and the parent state in this way vs having the parent be the source of truth and only using the props for current state?

eric-briscoe replied:
Dug into this more and was able to completely remove offline prop and state tracking with no affect to behavior

eschutho commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,118 @@
+      })
+        .then(({ json }) => {
+          if (json && Object.keys(json)?.length > 0) {
+            actions?.refreshQueries(json);
```
is there ever a case where you want to clear out all of the queries, e.g., the json comes back blank?

eric-briscoe replied:
I was not sure why it was originally written this way so I had left it alone.  I removed the check and found no side affect / change to behavior and updated to remove the Object.keys check

eschutho commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,118 @@
+        })
+        .catch(() => {
+          setIsOffline(true);
+          setPendingRequest(false);
```
nit, but maybe you can put setPendingRequests(false) into a finally event to dry it up a little.

eric-briscoe replied:
Great catch!   Will make change

eric-briscoe replied:
Updated in new commit

eschutho reviewed (commented):
Thanks @eric-briscoe, this is great! The comments make it very readable, too.

eric-briscoe committed:
addressing PR comments [sc-48362]

Removes unneeded props and state tracking of offline, adds finally block to simplify clearing pending request, simplifies value comparison in array by using includes in place of indexOf

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

AAfghahi commented on superset-frontend/src/SqlLab/fixtures.ts:
```diff
@@ -512,7 +513,86 @@ export const failedQueryWithErrors = {
 };
 
-export const runningQuery = {
+const baseQuery: Query = {
```
crazy that this hasn't existed yet!

AAfghahi reviewed (commented):
Looks good! Does it need an ephemeral for testing?

eric-briscoe commented on superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx:
```diff
@@ -0,0 +1,109 @@
+}
+
+// returns true if the Query.state matches one of the specifc values indicating the query is still processing on server
+export const isQueryRunning = (q: Query): boolean =>
```
@AAfghahi Agree it would be better to create a typescript enum for this in the types.ts file where Query is defined and use that. 

`/ / Possible states of a query object for processing on the server
export enum QueryState {
  STARTED = 'started',
  STOPPED = 'stopped',
  FAILED = 'failed',
  PENDING = 'pending',
  RUNNING = 'running',
  SCHEDULED = 'scheduled',
  SUCCESS = 'success',
  FETCHING = 'fetching',
  TIMED_OUT = 'timed_out',
}

// Indicates a Query's state is still processing
export const runningQueryStateList: QueryState[] = [
  QueryState.RUNNING,
  QueryState.STARTED,
  QueryState.PENDING,
  QueryState.FETCHING,
];

// Indictes a Query's state has completed processing regardless of success / failure
export const concludedQueryStateList: QueryState[] = [
  QueryState.STOPPED,
  QueryState.FAILED,
  QueryState.SUCCESS,
  QueryState.TIMED_OUT,
];

// check if a query is in a running state
export const isQueryRunning = (q: Query): boolean => runningQueryStateList.includes(q?.state);
`

 I have this setup and working on local but saw an additional query state value 'scheduled' in types.ts that seems like it should be included in the array used by isQueryRunning.  Anyone have context if 'scheduled' should be added to current list of ['running', 'started', 'pending', 'fetching']?

AAfghahi replied:
Oh yeah I remember this, it is a bit confusing because I think some of these states only occur in the frontend and never get passed to the backend. I once broke this in superset, let me check to see what I needed to do to fix

AAfghahi replied:
I think this is all of them: 
https://github.com/apache/superset/blob/master/superset-frontend/src/SqlLab/components/QueryTable/index.tsx#L117

eschutho replied:
So I looked into that a bit, and it's related to Scheduled Queries. The config/feature doesn't seem to be working as expected, but regardless, let's assume that it will create a query with a status of schedule, so I'd say yes, let's go ahead and add that value as an option.

eric-briscoe replied:
@AAfghahi @eschutho Thank you for the context.  I just pushed up a commit that adds QueryState.SCHEDULED to runningQueryStateList

eric-briscoe committed:
Address PR comment to use enum for QueryState [sc-48362]

Original implementation had string literals used in multiple places representing Query.state value options.  This commit creates a formal TypeScript enum for QueryState so we can remove string literals and ensure better consistency

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

eric-briscoe committed:
Address PR comments for object type validation [sc-48362]

This commit resolves an issue why the TypeScript typing for queryList was marked as a Query[] but was actually a Dictionary (associative array) or Queries.  A new type QueryDictionary has been added and the QueryAutoRefresh code was adjusted to use QueryDictionary instead of Query[] in appropriate places as well as unit tests.  Commit also removes QueryAutoRefreshContainer by making the once component using QueryAutoRefresh (which is already redux connected) pass the needed values on props.  this simplifies the code base and reduce files that need unit testing while keeping QueryAutoRefresh out of needing a redux connection directly.

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

eric-briscoe committed:
Addresses PR comment to add QueryState.SCHEDULED to runningQueryStateList [sc-48362]

In previous implementation 'scheduled' was not included int he list of Query States.  Further investigation shows it should be added to as a running state.

AAfghahi commented:
@eric-briscoe do you have the pre-commit hook installed? We have it in a makefile. The command is `make pre-commit`

eric-briscoe committed:
Fix prettier lint error [sc-48362]

eric-briscoe committed:
Adjust unit tests for props update hoisting callbacks out of actions wrapper object [sc-48362]

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

hughhhh commented on superset-frontend/src/SqlLab/utils/useInterval.ts:
```diff
@@ -0,0 +1,47 @@
+ * The useInterval function solves this issue.
+ * more info: https://overreacted.io/making-setinterval-declarative-with-react-hooks/
+ */
+function useInterval(callback: Function, delay: number | null): void {
```
nit: can we set delay as an optional and = to 0
```
  function useInterval(callback: Function, delay: number = 0): void {
```

eschutho replied:
@hughhhh I'm thinking it may be dangerous to run the function continuously without any delay interval (or the minimum interval that the browser allows). If you forget to pass a delay, this could be problematic. I think @eric-briscoe has a good idea about doing nothing (returning a noop). WDYT?

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

eric-briscoe committed:
Update with changes from master [sc-48362]

Merges in updates from master and resolves conflicts from relocation of some of the Query TypeScript definitions into core

eschutho reviewed (approved):
Thanks @eric-briscoe, looks great! There are just some tests and checks that aren't passing still.

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-48362/convert-queryautorefresh-to-functional-component

eric-briscoe committed:
Removes logic setting user offline and relying on results panel error message [sc-48362]

eric-briscoe committed:
Fixes bad import after some TypeScript definitions were relocated to core [sc-48362]

eric-briscoe committed:
Fixes TypeScript errors [sc-48362]
